src/metabase/lib/order_by.cljc

```diff
@@ -133,30 +133,37 @@
 
   ([query        :- ::lib.schema/query
     stage-number :- :int]
-   (let [indexed-order-bys (map-indexed (fn [pos [_tag _opts expr]]
-                                          [pos expr])
-                                        (order-bys query stage-number))
-         order-by-pos
-         (fn [x]
-           (some (fn [[pos existing-order-by]]
-                   (let [a-ref (lib.ref/ref x)]
-                     (when (or (lib.equality/= a-ref existing-order-by)
-                               (lib.equality/= a-ref (lib.util/with-default-effective-type existing-order-by)))
-                       pos)))
-                 indexed-order-bys))
-
-         breakouts          (not-empty (lib.breakout/breakouts-metadata query stage-number))
+   (let [breakouts          (not-empty (lib.breakout/breakouts-metadata query stage-number))
          aggregations       (not-empty (lib.aggregation/aggregations-metadata query stage-number))
          columns            (if (or breakouts aggregations)
                               (concat breakouts aggregations)
                               (let [stage (lib.util/query-stage query stage-number)]
-                                (lib.metadata.calculation/visible-columns query stage-number stage)))]
-     (some->> (not-empty columns)
-              (into [] (comp (filter orderable-column?)
-                             (map (fn [col]
-                                    (let [pos (order-by-pos col)]
-                                      (cond-> col
-                                        pos (assoc :order-by-position pos)))))))))))
+                                (lib.metadata.calculation/visible-columns query stage-number stage)))
+         columns            (filter orderable-column? columns)
+         existing-order-bys (order-bys query stage-number)]
+     (cond
+       (empty? columns)
+       nil
+
+       (empty? existing-order-bys)
+       (vec columns)
+
+       :else
+       (let [refs          (mapv lib.ref/ref columns)
+             ref->position (into {}
+                                 (map-indexed (fn [pos [_tag _opts expr]]
+                                                (when-let [closest-ref (lib.equality/find-closest-matching-ref
+                                                                        query
+                                                                        (lib.ref/ref expr)
+                                                                        refs)]
+                                                  [closest-ref pos])))
+                                 existing-order-bys)]
```

Comment:
This seems like an extractable helper - isn't there similar logic for other indexed things (like aggregations)?

Reply:
The aggregation logic is different, it's doing something weird with temporal units (as you've seen above), I did experiment with trying to abstract out more stuff but what I have in this PR ended up seeming to be the best fit for the current usages. We can always abstract out more stuff in the future if we need to I think